File: core/src/test/java/com/mower/domain/CommandShould.java
package com.mower.domain;

import org.junit.jupiter.api.Test;

import static com.mower.domain.CardinalPoint.NORTH;
import static com.mower.domain.Command.LEFT;
import static com.mower.domain.Command.MOVE;
import static com.mower.domain.Command.RIGHT;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

class CommandShould {

  private static final String UNKNOWN_COMMAND_CODE = "UNKNOWN COMMAND CODE";
  private static final int COORDINATE_X = 1;
  private static final int COORDINATE_Y = 1;
  private static final CardinalPoint CARDINAL_POINT = NORTH;

  @Test
  void throwUnknownCommandCodeIfCardinalPointCodeIsUnknown() {
    assertThrows(UnknownCommandCode.class, () -> Command.fromCode(UNKNOWN_COMMAND_CODE));
  }

  @Test
  void returnCommandLeftFromLCode() {
    assertSame(LEFT, Command.fromCode("L"));
  }

  @Test
  void returnCommandRightFromRCode() {
    assertSame(RIGHT, Command.fromCode("R"));
  }

  @Test
  void returnCommandMoveFromMCode() {
    assertSame(MOVE, Command.fromCode("M"));
  }

  @Test
  void executeLeftCommandAction() {
    LEFT.execute(mower());
  }

  @Test
  void executeRightCommandAction() {
    RIGHT.execute(mower());
  }

  @Test
  void executeMoveCommandAction() {
    MOVE.execute(mower());
  }

  private Mower mower() {
    return new Mower(coordinates(), faceTo());
  }

  private Coordinates coordinates() {
    return new Coordinates(COORDINATE_X, COORDINATE_Y);
  }

  private FaceTo faceTo() {
    return new FaceTo(CARDINAL_POINT);
  }
}

File: core/src/test/kotlin/com/mower/domain/CommandShould.kt
package com.mower.domain

import com.mower.domain.Command.Companion.fromCode
import com.mower.domain.Command.LEFT
import com.mower.domain.Command.MOVE
import com.mower.domain.Command.RIGHT
import com.mower.domain.exception.UnknownCommandCode
import com.mower.domain.valueobjects.Coordinates
import com.mower.domain.valueobjects.FaceTo
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith
import kotlin.test.assertSame

class CommandShould {
    private val UNKNOWN_COMMAND_CODE = "UNKNOWN COMMAND CODE"
    private val COORDINATE_X = 1
    private val COORDINATE_Y = 1
    private val CARDINAL_POINT = CardinalPoint.NORTH

    @Test
    fun throwUnknownCommandCodeIfCardinalPointCodeIsUnknown() {
        assertFailsWith<UnknownCommandCode> {
            fromCode(UNKNOWN_COMMAND_CODE)
        }
    }

    @Test
    fun returnCommandLeftFromLCode() {
        assertSame(LEFT, fromCode("L"))
    }

    @Test
    fun returnCommandRightFromRCode() {
        assertSame(RIGHT, fromCode("R"))
    }

    @Test
    fun returnCommandMoveFromMCode() {
        assertSame(MOVE, fromCode("M"))
    }

    @Test
    fun executeLeftCommandAction() {
        LEFT.execute(mower())
    }

    @Test
    fun executeRightCommandAction() {
        RIGHT.execute(mower())
    }

    @Test
    fun executeMoveCommandAction() {
        MOVE.execute(mower())
    }

    private fun mower(): Mower {
        return Mower(coordinates(), faceTo())
    }

    private fun coordinates(): Coordinates {
        return Coordinates(COORDINATE_X, COORDINATE_Y)
    }

    private fun faceTo(): FaceTo {
        return FaceTo(CARDINAL_POINT)
    }
}
